### ISSUE TITLE
Doesn’t record integrity hash for a remote tarball

### ISSUE THREAD

**Issue by andersk** [pnpm/pnpm#6328]
### pnpm version:

8.1.0

### Code to reproduce the issue:

`pnpm i https://registry.yarnpkg.com/lodash/-/lodash-4.17.21.tgz`

(Using `registry.yarnpkg.com` for this demonstration just so pnpm doesn’t detect it as the main NPM registry—any URL other than `registry.npmjs.org` has the same behavior.)

### Expected behavior:

pnpm should record an integrity hash for the tarball to maintain determinism and protect against potential MITM attacks, like it does for packages from the main NPM registry (`pnpm i lodash`).

### Actual behavior:

pnpm only records the tarball URL, with no integrity hash.

```console
$ pnpm i https://registry.yarnpkg.com/lodash/-/lodash-4.17.21.tgz
$ cat pnpm-lock.yaml 
lockfileVersion: '6.0'

dependencies:
  lodash:
    specifier: https://registry.yarnpkg.com/lodash/-/lodash-4.17.21.tgz
    version: '@registry.yarnpkg.com/lodash/-/lodash-4.17.21.tgz'

packages:

  '@registry.yarnpkg.com/lodash/-/lodash-4.17.21.tgz':
    resolution: {tarball: https://registry.yarnpkg.com/lodash/-/lodash-4.17.21.tgz}
    name: lodash
    version: 4.17.21
    dev: false
```

If I manually add an integrity hash to the lockfile, pnpm will check downloads against it, but it will remove the integrity hash whenever it rewrites the lockfile.

### Additional information:

 - `node -v` prints: v18.15.0
 - Windows, macOS, or Linux?: Linux

**Comment by capndave** [comment]
I've been looking at this a bit. I wonder if the integrity hash is something that the npm registry provides that wouldn't be available elsewhere. I'm still looking into it so will add more later. If you're a contributor or intimately familiar with this project, feel free to steer me in the right direction!

**Comment by andersk** [comment]
Yeah, the npm registry provides an integrity hash for packages in the registry, but we can also compute an integrity hash for a remote tarball when we download it (which we need to do anyway for a remote tarball in order to understand its dependencies). This is what npm and Yarn do.

**Comment by zkochan** [comment]
The problem with this is that if we don't know the integrity of the tarball before downloading it, then we don't really know if the downloaded tarball is valid or not.

Currently the integrity is always returned by the resolver. To do this, we'll have to do some refactoring because now the fetcher will have to return the integrity.

**Comment by andersk** [comment]
Obviously we can’t verify that the *first* download is valid when creating the lockfile. The point is to verify that we get the same result the *second* time when we already have the lockfile (even on a different system and network after some time has passed). That’s what lockfiles are for, after all.

**Comment by zkochan** [comment]
No, usually we know the right integrity on the first download because we get it from the package's metadata.

**Comment by andersk** [comment]
Yes I understand the mechanics; I said that myself above. But my concern as a user is not that the package is consistent with its own metadata—my concern is that my production deployment is consistent with the version I tested locally and in CI.

For packages in the registry, without the lockfile, I might get a different version of a dependency as updates are released over time. With lockfile support, I might still get a bad version the first time (“bad” in the sense that some bug in that version affects me, not necessarily in the sense that the download was corrupted), but I can be confident if I get a good version the first time, it won’t silently give me a bad version the second time.

Similarly, for remote tarballs on arbitrary servers, without the integrity hash, there’s nothing to guard against the tarball being replaced on the server with different code. Other package managers check this, and pnpm should too.

**Comment by andrewhamon** [comment]
> No, usually we know the right integrity on the first download because we get it from the package's metadata.

Both of these mechanisms are Trust On First Use. Pnpm is still trusting the server to serve up authentic metadata on first use. The security properties of either mechanism are very similar.

**Comment by lougreenwood** [comment]
I think that this also causes issues with dependencies coming from github causing `pnpm licenses` to fail:

https://github.com/pnpm/pnpm/issues/8739

I did some research and it seems that `pnpm licenses` will only process dependencies [which have a value for integrity](https://github.com/pnpm/pnpm/blob/8c3de19fc11ad000f778027f9dcf7a349e4a3181/reviewing/license-scanner/src/getPkgInfo.ts#L255)... and when I check in my lock file I see that the github depenendency is also missing an integrity value.

**Comment by andersk** [comment]
It looks like this was fixed in 11.9.0:

- #12491
